### PR TITLE
fix: Add support for Wayland window decorations

### LINKED ISSUE #27522: [Feature Request] Support CSD in Wayland
**Preflight Checklist**

* [x] I have read the [Contributing Guidelines](https://github.com/electron/electron/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [Code of Conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

**Problem Description**

In Wayland clients usually render their own decorations, they don't get SSD anymore (although there seems to be a protocol for SSD negotiation that Plasma and some compositors based on wlroots implement). The Wayland build is not very useful until Electron app windows get a proper headerbar. The headerbar can be a stock headerbar similar to old-school SSD or can be customized by the app (for example, by setting its color/theme in accordance with the app contents). 

**Proposed Solution**

Use GTK CSD support, which is also well-integrated in Qt-based environments (mainly Plasma):

https://developer.gnome.org/gtk3/stable/GtkHeaderBar.html:

```
void gtk_header_bar_set_show_close_button (GtkHeaderBar *bar, gboolean setting);

Sets whether this header bar shows the standard window decorations, including close, maximize, and minimize.
```

https://developer.gnome.org/gtk3/stable/GtkSettings.html#GtkSettings--gtk-decoration-layout:

```
“gtk-decoration-layout”    char *

This setting determines which buttons should be put in the titlebar of client-side decorated windows, and whether they should be placed at the left of right.
```

**Alternatives Considered**

It doesn't seem to exist much more alternatives. Maybe Qt has a native headerbar implementation, but it won't properly work inside a GTK-based environment like GNOME, while it works the other way around. See https://pointieststick.com/2019/11/30/this-week-in-kde-gtk-csd-support-and-more

Maybe xdg-decoration as suggested in some comments below. But I ignore everything about it and SSD doesn't seem to me the best alternative for Electron apps that tend to be more app-centric than platform-centric by its very nature and so, at least in major platforms, usually customize headerbars.

**Additional Information**

More details in discussion starting from https://github.com/electron/electron/issues/10915#issuecomment-767997150

### PR DESCRIPTION
**Pull request by refi64**
#### Description of Change

This implements client-side-drawn window decorations for native Wayland usage. Closes #27522.

Current notes (reasons for WIP mark):

- ~~I have not tested this with many GTK themes or on many compositors~~ Has been now tested with Yaru, Pop's theme, and a few others (CrosAdapta had some glitches but upstream Chromium breaks under it too, so it's not just on us). In addition, on Sway it has been tested to continue using SSD. 
- ~~Menus are, uhh, offset from their correct position? Not clear if this is an issue related to upstream Wayland support...but realistically it's probably on my end.~~ Fixed now. This may also implement drag-and-drop...but right now many drag-and-drop uses on Chromium Wayland instantly crash the whole thing anyway, so I can't even test that part yet.
- I have also not tested this with a particularly large variety of applications yet.
- Wasn't sure what to put under the per-file "copyright" section, please point out if what I put there is incorrect?

Demo:

https://user-images.githubusercontent.com/1690697/121403756-dc793700-c920-11eb-9e68-36e251415585.mp4

#### Checklist

- [x] PR description included and stakeholders cc'd (@ckerr @zcbenz, apologies in advance if this isn't correct, but I had no idea how to find people and just copied the cc's from #26022 :sweat_smile:)
- [x] `npm test` passes
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: none

### CODE REVIEW

**Comment by refi64** [comment]
Right, almost forgot: there are also 2 TODOs in the source, related to an accessibility string define that's present in Chromium but missing here + `GetStringUTF16` crashing, does Electron..."trim" out accessibility strings from Chromium that it doesn't use?

**Comment by vially** [comment]
I think this is looking great!

I'm planning to build and test this locally but until I get around to do that I was wondering how is this going to affect compositors that support [server-side decorations](https://wayland.app/protocols/xdg-decoration-unstable-v1) (e.g.: Plasma, sway, etc)?

Are they going to get duplicate decorations (one from the compositor and one from the client)?

**Comment by refi64** [comment]
I believe the server-side decoration protocol [requires the client to essentially opt-in](https://wayland.app/protocols/xdg-decoration-unstable-v1):

> If compositor and client do not negotiate the use of a server-side decoration using this protocol, clients continue to self-decorate as they see fit.

From some stepping through the source code, it seems like this only occurs if `remove_standard_frame` is set to `false`. However, this PR ends up setting it to `true` unconditionally, thus as-is this should work "fine". That being said, losing the SSD isn't ideal, so I'll take a look into seeing if we can only enable CSD depending on whether or not a native frame is used.

**Comment by refi64** [comment]
@vially as of the latest commit, it should go back to using SSD on compositors that support it. I couldn't test with KDE, but I *did* test with Sway, where it appeared to work.

**Comment by khimaros** [comment]
following along with excitement. thank you for taking on this complex task @refi64!

**Comment by refi64** [comment]
From some local testing, I believe this should be usable enough for general use. There are two outstanding open notes:

- The first Wayland patch isn't incredibly large, but it *has* already hit a merge conflict due to a rather unlucky renaming that impacted one of the few functions the patch uses. This would be something that would ideally be merged upstream, but I'm not sure if Chromium is generally open to accepting code it doesn't need *yet*?
- Related to the above, I was unable to test cross-window drag-and-drop, as it crashed an entirely unrelated section of the Wayland backend.

I believe these issues are not "blocking" per se because:

- The Wayland backend itself is still unstable and does hit upstream bugs from time to time.
- Right now, there are no window decorations on GNOME Wayland, so having *something* is still a sizable improvement.
- Note that the test failures seem to be entirely unrelated.

With this in mind, I'm now marking this PR as ready for review.

**Comment by major-mayer** [comment]
Thank you very much for that PR :heart:
I haven't though that someone would really make the effort to patch Electron just to support CSD on one single compositor under Wayland.
AFAIK all other compositors apart from Gnome support SSD under Wayland.

As a Gnome on Wayland user i would love to see this getting merged fast, because using Electron apps without any window decorations is "somewhat less than perfect".

**Comment by mystiquewolf** [comment]
@refi64 Please take a look at #29523, looks related.

**Comment by ZanderBrown** [comment]
> I haven't though that someone would really make the effort to patch Electron just to support CSD on one single compositor 

Well GNOME/Mutter isn't exactly obscure

> AFAIK all other compositors apart from Gnome support SSD under Wayland.

Weston, the reference implementation, doesn't - even those that support `zxdg_decoration` may not always use SSD (if ever)

>  I'm not a desktop dev by any means, but did you need to introduce any GTK related dependencies to make this work?

Electron already depends on Gtk

> IIRC this was the reason why the Mutter devs rejected SSD in the first place

Mutter already depends on Gtk

**Comment by major-mayer** [comment]
> Mutter already depends on Gtk

Maybe I'm wrong about the other points, but at least this one could be right, as explained in this comment :https://gitlab.gnome.org/GNOME/mutter/-/issues/217#note_357576 

However this is quite offtopic, so nevermind.

File: patches/chromium/ozone_wayland_add_support_for_content_insets.patch
Comment (zcbenz) [inline patches/chromium/ozone_wayland_add_support_for_content_insets.patch:146]: Why is Chromium not setting content insets itself? Do PWA apps of Chrome browser work correctly under Wayland CSD?

Comment (refi64) [reply patches/chromium/ozone_wayland_add_support_for_content_insets.patch:146]: @zcbenz Yes but actually no...it depends on your definition of "work correctly".

If you're on Xorg and using CSD, Chromium does not use any shadows at all, instead opting for a rather large and [incredibly unpopular 3D border effect](https://bugs.chromium.org/p/chromium/issues/detail?id=920768) to "mimic" the look. If you're on Wayland, and you switch the window decorations to look native, the title bar has no borders or shadows at all and also looks rather terrible.

As of right now, I am not sure if this will ever be fixed upstream. I had hopes that LaCrOS would result in this being fixed...but it appears that they've just opted for the shadows to be provided by the Chrome OS compositor rather than done by the client, unfortunately, meaning there's no currently available workaround for native Linux.

**Review by nornagon — CHANGES_REQUESTED** [review]
This is quite a large and fairly intricate patch. What are the obstacles to upstreaming this? I'd prefer not to take on a "forever patch" of this scale.

**Comment by refi64** [comment]
@nornagon does Chromium normally take patches for things not used in Chromium itself? I believe upstreaming it is probably doable otherwise, but I'd expect "we're not actively using this" to be a roadblock?

**Comment by nornagon** [comment]
@refi64 It depends, but broadly speaking, no. It's worth trying anyway though, as sometimes you'll get a friendly reviewer and/or you can make a case for why it should be merged anyway (e.g. chromium might use it in the future; it sounds like there is a potential path there based on an argument that their window decorations should be better on wayland?)

If upstreaming this patch is a complete dead end, we should explore, in order of preference:

1. Can we implement this functionality without patching? e.g. can we subclass, or implement delegate methods, or fetch the underlying handles, etc. to avoid a patch?
2. Can we upstream a patch that refactors the code in an acceptable-to-upstream way that allows us to do (1)?
3. Can we make this patch smaller?

If we can't do any of the above, I am 👎  on this PR. Every patch we take on creates a significant maintenance burden. If landed as-is, this patch would be the largest Chromium patch we are carrying, by a factor of 2. A patch of this size is not worth the ongoing maintenance cost for what it buys, IMO.

**Comment by refi64** [comment]
Understood, I'll start taking a look into what it would take to upstream this. There are a few patches we're carrying in the Chromium Flatpak I should be upstreaming soon too, I guess this is extra incentive!

**Comment by refi64** [comment]
@nornagon as another thought, would a version of this PR with all support for shadows removed be accepted? That would probably still be better than the current state, at least until the related patches can be upstreamed, then the shadow support could be added back here.

**Comment by nornagon** [comment]
@refi64 If that meant a significantly smaller Chromium patch (or ideally no patch at all), that would probably be worth exploring.

**Comment by zcbenz** [comment]
> If you're on Xorg and using CSD, Chromium does not use any shadows at all, instead opting for a rather large and [incredibly unpopular 3D border effect](https://bugs.chromium.org/p/chromium/issues/detail?id=920768) to "mimic" the look. If you're on Wayland, and you switch the window decorations to look native, the title bar has no borders or shadows at all and also looks rather terrible.

@refi64 Would it be feasible to reuse your code in Chromium to improve the Chrome browser's experience with Wayland CSD? I think that would might be more persuasive to make Chromium accept your patch, though it might need much more work.

**Comment by ckerr** [comment]
@refi64 Thank you for being understanding about trying to upstream this into Chromium or otherwise finding a way to eliminate or minimize the footprint of the diff.

The issue is that upstream Chromium code gets refactored pretty aggressively. On any given day we literally have at least one engineer whose primary responsibility is keeping `main` up-to-date with upstream changes. One of the main timesinks is it's _extremely_ common for our patches to break due to upstream changes. We typically avoid _any_ patch that doesn't have some kind of endgame if we can avoid it.That's why the the maintenance cost of a large forever-patch can be very expensive over time.

That said, I hope we can find some way of upstreaming this or shrinking it. I love the feature and would use it myself. :heart:

**Comment by refi64** [comment]
@nornagon @zcbenz @ckerr so the latest version should have *no* Chromium patches at all. It turns out, the transparency fix was no longer needed for this directly, but it still is required for transparency to work, so I pulled it out, and I'll submit it in a separate PR.

File: shell/browser/native_window.h
Comment (zcbenz) [inline shell/browser/native_window.h:318]: This method is not used anywhere.

**Review by zcbenz — APPROVED** [review]
The implementation looks solid to me.

File: shell/browser/native_window_views.cc
Comment (nornagon) [inline shell/browser/native_window_views.cc:222]: This change looks like it might have effects beyond the description of the PR—does this fix a bug about resizability on Linux by any chance?

File: shell/browser/ui/views/client_frame_view_linux.cc
Comment (nornagon) [inline shell/browser/ui/views/client_frame_view_linux.cc:42]:
```suggestion
// The border color is encoded in box-shadow, which also cannot be read, so
```

File: shell/browser/ui/views/client_frame_view_linux.cc
Comment (nornagon) [inline shell/browser/ui/views/client_frame_view_linux.cc:39]: > for now

Is this expected to be readable from the system theme somehow in future? I'm not familiar with wayland at all so forgive me if this is a nonsensical question 😅

File: shell/browser/ui/views/client_frame_view_linux.cc
Comment (nornagon) [inline shell/browser/ui/views/client_frame_view_linux.cc:115]: Are we sure `base::Unretained` is OK here? If so, let's add a short comment here about the reason why it's OK.

File: shell/browser/ui/views/client_frame_view_linux.cc
Comment (nornagon) [inline shell/browser/ui/views/client_frame_view_linux.cc:121]: Here too.

**Review by nornagon — APPROVED** [review]

File: shell/browser/ui/views/client_frame_view_linux.cc
Comment (refi64) [reply shell/browser/ui/views/client_frame_view_linux.cc]: Good point, this is entirely dependent on GTK3 and I don't think they plan on adding an official API for this.

File: shell/browser/native_window_views.cc
Comment (refi64) [reply shell/browser/native_window_views.cc:222]: Hmm that's true, this would in theory affect `resizable: true` on any FramelessView, right? I can't find any bug reports on that though, maybe it's just a rarely-used combination?

File: shell/browser/ui/views/client_frame_view_linux.cc
Comment (nornagon) [reply shell/browser/ui/views/client_frame_view_linux.cc]: Hm, does this mean we'll be locked into not matching the system theme in some cases? Seems like a poor design choice on Wayland's part if that's so, as it doesn't seem possible to blend in.

Comment (refi64) [reply shell/browser/ui/views/client_frame_view_linux.cc]: It's definitely rather unfortunate, though it's more on GTK3's part than Wayland's part. There may be increasingly hack-ish ways to derive this value, but I'm just not sure if it would really be worth it: as-is the current choice does seem to blend with the theme choices on the most popular distro variants (matches with the GNOME default Adwaita used on Arch, and looks fine with Yaru used on Ubuntu, I believe the only major distros that ship a theme that conflict would be Pop and *maybe* Manjaro?).

I believe it's worth noting that I believe Chromium itself will just always use a small border radius, though I fear that just using a smaller one will just make it look out of place on all systems.

**Comment by refi64** [comment]
CI issues again seem to be unrelated to this change.

**Comment by refi64** [comment]
So...it looks like Chromium is doing its [own updated CSD work](https://chromium.googlesource.com/chromium/src/+/e69e819f2629437bee5f3d95f24fb3cbc2013fbc):

> A followup will use GTK to render a native frame (including rounded
corners, border, and shadow) when the GTK theme is used.

Thus it might be best to hold off on this for a bit until this new work is merged in upstream.

**Comment by ckerr** [comment]
@refi64 if you're going to monitor that upstream, could you @ me in this PR when there's an update?

As a procedural change, I'm going to mark this PR as a draft so that it doesn't get merged accidentally if another maintainer doesn't read all the way down into the comments. @refi64 feel free to change it back as needed.

**Comment by eyelash** [comment]
[The work to render native window frames using GTK in Chromium](https://chromium-review.googlesource.com/c/chromium/src/+/3033187) seems to be merged now.

**Comment by refi64** [comment]
FYI I'll be working on updating this *very* soon, there was some hardware-related stuff that's just been in flux and messed with by planning for a bit.

**Comment by GreatGodvin** [comment]
Would using [libdecor](https://gitlab.gnome.org/jadahl/libdecor) make things easier ?

**Comment by iMonZ** [comment]
Anything new here?

**Comment by major-mayer** [comment]
I've just installed Chromium 96 to test out the new GTK based window decorations and they work like a charm.
Can't wait to see this also happening in Electron:
image

**Comment by major-mayer** [comment]
Any updates when this could get merged?

**Comment by refi64** [comment]
After a decent bit of work, this has been redone on top of the new native decoration support, meaning we get shadows w/o needing any patches (yay!) and don't have any issues with a hardcoded border radius. (ping @ckerr)

**Review by zcbenz — APPROVED** [review]
New code looks good to me.

**Comment by Devorlon** [comment]
kermit-flail

**Comment by DiegoMagdaleno** [comment]
The new code is great and it provides a better UX for all wayland users, including those that arent directly in gnome.

For example wayfire and sway.

And in the future XFCE plans to migrate to mutter, so in my opinion this is gonna be necessary for electron apps to provide the best UX in the GTK desktops.

Thank you for your work, fingers crossed it gets merged.

**Comment by eyelash** [comment]
@refi64 Could this code also be used to provide more native CSD on X11? If I turn on CSD in VSCode for example the window border and the window buttons look different from other apps while on Windows and macOS they look native.

**Comment by refi64** [comment]
@eyelash I believe the window buttons would be the responsibility of VSCode itself. As for the borders, this code could indeed be used to draw borders & shadows, but I'm not sure about all the details. (I know `frameless: true` will remove the title bar, but I'm not sure if it preserves borders & shadows outside of Wayland; if it does, then a future PR could be sent based on this one that would draw the borders and shadows when `frameless: true`.)

**Comment by eyelash** [comment]
I'm not an electron developer myself (only a user of electron apps) so I might be wrong but I found [this documentation](https://www.electronjs.org/docs/latest/tutorial/window-customization) that talks about `titleBarStyle: 'hidden'` so I'm assuming electron has the ability to draw the window buttons on Windows and macOS:

> Title bar styles allow you to hide most of a BrowserWindow's chrome while keeping the system's native window controls intact

**Comment by rohmishra** [comment]
@eyelash 

The CSD thing on windows with custom CSD is actually a hack since it draws the buttons on its own just like it does on linux. 
Windows has only one button configuration so you dont have to worry about anything else and can recreate it to look as native as possible.

With macOS, its the same thing. AFAIK, apple's API actually provide a way to place the traffic lights as needed. Im not sure if VSCode uses that, but even if it doesn't - recreating them is rather simple since you have only one configuration.

What makes linux systems tough is that user/distro/DM has the choice on which button to keep and where to place them.

I dont have maximize/restore option on my system for example and have the buttons on left. Elementary has split their buttons to be on both sides. KDE by default does three buttons on right like windows. Gnome does two buttons, also on right. Someone might choose to keep only the close button and discard everything.

There is no standard way to read this configuration as all DM have their own way to keep track of this. Plus apps cant possibly test against all themes out there that have different button glyphs. A compromise would be exposing button configuration to app and letting them use their own glyphs.

**Comment by ZanderBrown** [comment]
>  Gnome does two buttons

You mean one

> There is no standard way to read this configuration

Just use GtkWindowControls, still a little more complex than macOS (where of course it's one widget on the left) since you need two (one on each side)

The more complex question (which would affect win32/macOS as well) is how apps know the "safe area"

**Comment by rohmishra** [comment]
Doesnt GtkWindowControls require Gtk4 @ZanderBrown ? Also does it have a property to force color scheme since app may have their UI bright requiring dark colored buttons.

**Comment by ZanderBrown** [comment]
> Doesnt GtkWindowControls require Gtk4

Yes and no, it's true gtk3 doesn't include it, but it's simple enough to do in gtk3 (it was originally prototyped in gtk3) so could be kept in-tree until chromium+gtk4 happens (which may be sooner than you'd think anyway)

> does it have a property to force color scheme since app may have their UI bright requiring dark colored buttons.

Huh? You seem to have flipped something there, and I'd argue electron apps being able to hint their light/dark scheme is a wider issue

**Comment by rohmishra** [comment]
Hmm, I see.

Yes I was talking about apps hinting color scheme/requesting gtk elements to be of a specific color scheme. AFAIK there isn't any mechanism to do that yet.

**Comment by alvarlagerlof** [comment]
Is there anything else needed for Electron to work good on wayland?

**Comment by DiegoMagdaleno** [comment]
@alvarlagerlof Doesn't look like it, but it doesn't look like devs are planning on merging this either

**Comment by csett86** [comment]
Its quiet month (https://www.electronjs.org/blog/a-quiet-place-21) in December this year for the project, thus no merging of PRs, @DiegoMagdaleno 

But I would not be worried, the pr is already approved.

**Comment by alvarlagerlof** [comment]
Ah I see, that makes sense. Thanks @csett86

**Comment by clydebottoms** [comment]
can this be merged now?

**Comment by nornagon** [comment]
I'm fine with this, but I'd like someone on the maintainer team who is more familiar with Wayland and Linux to sign off before merging. @ckerr?

**Comment by DiegoMagdaleno** [comment]
Tbh, I wish this could be merged, almost every week, you see more and more Electron repos point back to this issue, as the Linux community moves towards Wayland, having proper Wayland support on a framework/runtime that is supposed to fix the desktop fragmentation problem would be great.

And im scared this becomes one of the PRs that never gets merged and is just referenced as "There was an attempt at fixing this issue in [link] but it was never upstreamed"

**Comment by iMonZ** [comment]
> I'm fine with this, but I'd like someone on the maintainer team who is more familiar with Wayland and Linux to sign off before merging. @ckerr?

Is this going to be merged at version 16 or 17?

**Comment by Yajo** [comment]
Can't be merged behind some experimental flag for now?

**Comment by nornagon** [comment]
> Can't be merged behind some experimental flag for now?

Actually yeah, I think we could do that, & it's a good idea. I'll give it a go.

File: shell/browser/ui/electron_desktop_window_tree_host_linux.h
Comment (refi64) [reply shell/browser/ui/electron_desktop_window_tree_host_linux.h]: (Out of curiosity, why the change? These are only ever used inside the class.)

Comment (nornagon) [reply shell/browser/ui/electron_desktop_window_tree_host_linux.h]: Style guide says always use `public`: https://google.github.io/styleguide/cppguide.html#Inheritance

**Comment by nornagon** [comment]
@refi64 I've added code to guard this behind `--enable-features=WaylandWindowDecorations` (should also work with `app.commandLine.appendSwitch`). Would you be able to try it out & let me know if it works for you?

**Comment by refi64** [comment]
I can try it out in a bit (building Electron is not very fast :sweat_smile:), however, will there be anything to track upgrading it from a feature flag to enabled-by-default? I do have a bit of a concern that it will just end up stuck as a flag for too long because it's forgotten about...

(For the record, do also remember that Wayland support itself is not enabled by default yet, so this code path wouldn't be taken unless the user was already overriding the Ozone platform selection.)

**Comment by rohmishra** [comment]
@refi64 it may have issues that prevent people from using eletron apps in with wayland. Some apps provide CSD buttons and are quite functional on wayland. Not sure about it though.

**Comment by DiegoMagdaleno** [comment]
I think the approach that @nornagon suggests is better than forcing GTK CSDs, mainly because, as mentioned, some apps already draw their own decorations, and it looks repetitive.

It is best if maintainers have to opt in to add this feature, or if we as users are able to pass the switch, tools like Flatpak and what not already have tools like Flatseal to manage this via a GUI.

But both approaches are fine in my opinion.

**Comment by refi64** [comment]
This code path is not taken (or at least, should not be taken, I tested it though) if the app requests to use the frameless view, so there will not be double window decorations.

> It is best if maintainers have to opt in to add this feature

Given that it's often a struggle to get many apps to support Linux well in the first place, asking every single app to flip this on isn't really viable.

> or if we as users are able to pass the switch, tools like Flatpak and what not already have tools like Flatseal to manage this via a GUI.

This wouldn't work for the case of native packages, and Flatpak overrides cannot add command line options anyway, so this isn't viable either.

**Comment by juxuanu** [comment]
It would work with env variables. Something like ELECTRON_WINDOW_DECORATIONS=1 can be set in Flatseal or globally to the whole system and unset per-app.

**Comment by LasseRosenow** [comment]
> It would work with env variables. Something like ELECTRON_WINDOW_DECORATIONS=1 can be set in Flatseal or globally to the whole system and unset per-app.

This is no solution in the long term. We can't expect users to enable system wide flags just so that their electron apps work on linux :(

**Comment by ZanderBrown** [comment]
> This is no solution in the long term. We can't expect users to enable system wide flags just so that their electron apps work on linux :(

A solution to what?

**Comment by Devorlon** [comment]
> A solution to what?

Whether electron draws a generic CSD or the application specifies its own.

**Comment by ZanderBrown** [comment]
Huh? Your just inventing problems there :-)

A generic electron window has a titlebar, the problem is that on the Wayland backend nobody implemented it — contrary to the expected, and documented, behaviour (i.e. that non-frameless windows are "framed")

For applications that wish to provide their own decorations (or "frame") they are free to request a frameless window

**Comment by rohmishra** [comment]
@Devorlon It seems like this codepath will only be taken IF the application is running wayland AND expects window decoration to be handled for it.
So apps like vscode (with custom titlebar option) wont be affected even if running under Wayland.

File: shell/browser/native_window.cc
Comment (arslan-raza-143) [reply shell/browser/native_window.cc]: > ```suggestion
> 
> ```

Comment (arslan-raza-143) [reply shell/browser/native_window.cc]: ``

**Comment by nornagon** [comment]
> building Electron is not very fast

btw you can download the CI artifacts and test them locally; no need to build your own :)

e.g. https://app.circleci.com/pipelines/github/electron/electron/48129/workflows/a70541a7-e0a5-45c5-8648-5d36fd73fd83/jobs/1092704/artifacts, download 'dist.zip'

**Comment by mitchchn** [comment]
I attempted to test this branch but got the same segfault in Wayland mode as on `main` unless I comment out the entire window state watcher (described in #18943). @refi64 does the CI artifact run for you with `--ozone-platform=wayland`?)

**Comment by major-mayer** [comment]
I've just tried this build artifact (started it with `./electron --ozone-platform=wayland --enable-features=WaylandWindowDecorations`) and for me it works :heart: 

grafik

I would love to see this getting merged now.

**Comment by jplatte** [comment]
Still crashes with the provided build for me, regardless of whether I pass `--enable-features=WaylandWindowDecorations`, on sway 1.6.1:

```
[63412:0126/131529.322436:ERROR:power_monitor_device_source_stub.cc(11)] Not implemented reached in virtual bool base::PowerMonitorDeviceSource::IsOnBatteryPower()

Electron 18.0.0-nightly.20220106 - Build cross platform desktop apps with JavaScript, HTML, and CSS
Usage: electron [options] [path]

A path to an Electron app may be specified. It must be one of the following:
  - index.js file.
  - Folder containing a package.json file.
  - Folder containing an index.js file.
  - .html/.htm file.
  - http://, https://, or file:// URL.

Options:
  -i, --interactive     Open a REPL to the main process.
  -r, --require         Module to preload (option can be repeated).
  -v, --version         Print the version.
  -a, --abi             Print the Node ABI version.
[63412:0126/131529.496807:ERROR:wayland_connection.cc(612)] Not implemented reached in static void ui::WaylandConnection::Global(void *, wl_registry *, uint32_t, const char *, uint32_t)
[63412:0126/131529.570432:ERROR:cursor_loader.cc(114)] Failed to load a platform cursor of type kNull
[63412:0126/131529.570863:ERROR:wayland_window.cc(393)] Not implemented reached in virtual void ui::WaylandWindow::SetWindowIcons(const gfx::ImageSkia &, const gfx::ImageSkia &)
Received signal 11 SEGV_MAPERR 000000000020
#0 0x5577db20fd09 base::debug::CollectStackTrace()
#1 0x5577db140633 base::debug::StackTrace::StackTrace()
#2 0x5577db20f791 base::debug::(anonymous namespace)::StackDumpSignalHandler()
#3 0x7fead5050870 (/usr/lib/libpthread-2.33.so+0x1386f)
#4 0x5577d72e3000 base::ObserverList<>::AddObserver()
#5 0x5577d72e8024 electron::WindowStateWatcher::WindowStateWatcher()
#6 0x5577d72d7342 electron::NativeWindowViews::NativeWindowViews()
#7 0x5577d72dab1b electron::NativeWindow::Create()
#8 0x5577d712ffc0 electron::api::BaseWindow::BaseWindow()
#9 0x5577d7143415 electron::api::BrowserWindow::BrowserWindow()
#10 0x5577d7145d1d electron::api::BrowserWindow::New()
#11 0x5577d7146255 gin_helper::internal::InvokeNew<>()
#12 0x5577d7136e8f gin_helper::Dispatcher<>::DispatchToCallback()
#13 0x5577d895cc8a v8::internal::FunctionCallbackArguments::Call()
#14 0x5577d8959e7a v8::internal::(anonymous namespace)::HandleApiCallHelper<>()
#15 0x5577d895868a v8::internal::Builtin_Impl_HandleApiCall()
#16 0x5577d89581bd v8::internal::Builtin_HandleApiCall()
#17 0x3b0607edc4b8 <unknown>
  r8: 00000e48002e5403  r9: 00000001ffffffff r10: 00000e4801201a30 r11: 0000000000000000
 r12: 00000e4800336200 r13: 00007ffdf0bca080 r14: 00000e4800352500 r15: 00000e4800336258
  di: 00000e4800352500  si: 00000e480141de40  bp: 00007ffdf0bc9ac0  bx: 00000e480141de40
  dx: 00007fead50590a0  ax: 0000000000000020  cx: 0000000000000000  sp: 00007ffdf0bc92b0
  ip: 00005577d72e3000 efl: 0000000000010202 cgf: 002b000000000033 erf: 0000000000000004
 trp: 000000000000000e msk: 0000000000000000 cr2: 0000000000000020
[end of stack trace]
[63450:0100/000000.600984:ERROR:gpu_init.cc(454)] Passthrough is not supported, GL is egl, ANGLE is
[1]    63412 segmentation fault (core dumped)  ./electron --ozone-platform=wayland
```

**Comment by nornagon** [comment]
Okay, it seems like at least _something_ is working with this patch, and it's not breaking non-wayland cases, so I'm going to go ahead and merge this. Hopefully we can address the crashes and eventually pull this out from behind a flag to make it the default behavior.

**Comment by refi64** [comment]
FWIW the crashes are not tied to this at all, I believe it was this bug: https://bugs.chromium.org/p/chromium/issues/detail?id=1279574&q=wlroots&can=1

**Comment by iMonZ** [comment]
> Okay, it seems like at least _something_ is working with this patch, and it's not breaking non-wayland cases, so I'm going to go ahead and merge this. Hopefully we can address the crashes and eventually pull this out from behind a flag to make it the default behavior.

To which version it is merged?

**Comment by iMonZ** [comment]
> Okay, it seems like at least _something_ is working with this patch, and it's not breaking non-wayland cases, so I'm going to go ahead and merge this. Hopefully we can address the crashes and eventually pull this out from behind a flag to make it the default behavior.

Is it possible to merge this with older version of electron as well.
Since it won’t break anything and later the flag can be removed if the bugs are fixed and the removal of the parameter can be done in a newer version of electron. Sounds like a good deal right?

**Comment by nornagon** [comment]
I'm happy to backport to beta (17.x) but I don't want to risk destabilizing 16.x and prior without further testing.

**Comment by iMonZ** [comment]
> I'm happy to backport to beta (17.x) but I don't want to risk destabilizing 16.x and prior without further testing.

Thank you so much!
Is it possible to backport it to version 16 as well with enough testing?
At the end, it still needs a flag to be enabled 🙃

**Comment by refi64** [comment]
Great to see this merged! Just for some clarity though, what's going to be the policy for promoting it from behind a flag?

**Comment by Jmennius** [comment]
I am trying this out with Slack, which currently runs on electron 18.1.0 from what I can tell.
For me, Slack window does not take all space when maximised - it leaves quite significant margins all around (18-21 px). It is fine when tiling to the left or right.
This is on GNOME 42.2, Fedora 36. I've tried disabling dash-to-panel which I use.

**Pull request merged**
